test/topology/test_tablets.py

```diff
@@ -34,34 +34,36 @@ async def test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode(m
     # s0 should miss schema and tablet changes
     await manager.server_stop_gracefully(s0)
 
-    manager.cql.execute("CREATE KEYSPACE test WITH replication = {'class': 'NetworkTopologyStrategy', "
+    cql = manager.get_cql()
+    await cql.run_async("CREATE KEYSPACE test WITH replication = {'class': 'NetworkTopologyStrategy', "
                         "'replication_factor': 3, 'initial_tablets': 100};")
 
     # force s0 to catch up later from the snapshot and not the raft log
     await inject_error_one_shot_on(manager, 'raft_server_force_snapshot', not_s0)
-    manager.cql.execute("CREATE TABLE test.test (pk int PRIMARY KEY, c int);")
+    await cql.run_async("CREATE TABLE test.test (pk int PRIMARY KEY, c int);")
 
     keys = range(10)
-    await asyncio.gather(*[manager.cql.run_async(f"INSERT INTO test.test (pk, c) VALUES ({k}, 1);") for k in keys])
+    await asyncio.gather(*[cql.run_async(f"INSERT INTO test.test (pk, c) VALUES ({k}, 1);") for k in keys])
 
-    rows = manager.cql.execute("SELECT * FROM test.test;")
+    rows = await cql.run_async("SELECT * FROM test.test;")
     assert len(list(rows)) == len(keys)
     for r in rows:
         assert r.c == 1
 
     manager.driver_close()
     await manager.server_start(s0, wait_others=2)
     await manager.driver_connect(server=servers[0])
-    await wait_for_cql_and_get_hosts(manager.cql, [servers[0]], time.time() + 60)
+    cql = manager.get_cql()
+    await wait_for_cql_and_get_hosts(cql, [servers[0]], time.time() + 60)
 
     # Trigger a schema change to invoke schema agreement waiting to make sure that s0 has the latest schema
-    manager.cql.execute("CREATE KEYSPACE test_dummy WITH replication = {'class': 'NetworkTopologyStrategy', "
+    await cql.run_async("CREATE KEYSPACE test_dummy WITH replication = {'class': 'NetworkTopologyStrategy', "
                         "'replication_factor': 1, 'initial_tablets': 1};")
 
-    await asyncio.gather(*[manager.cql.run_async(f"INSERT INTO test.test (pk, c) VALUES ({k}, 2);", execution_profile='whitelist')
+    await asyncio.gather(*[cql.run_async(f"INSERT INTO test.test (pk, c) VALUES ({k}, 2);", execution_profile='whitelist')
                            for k in keys])
 
-    rows = manager.cql.execute("SELECT * FROM test.test;").all()
+    rows = await cql.run_async("SELECT * FROM test.test;")
     assert len(rows) == len(keys)
```

Comment:
If you drop all(), doesn't it mean that after the first iteration over the (paginated) result set in `assert len(rows)`, the second loop `for r in rows` will be no-op? See c39332710d3d0e4e17c879ea0b58550fecd12518

Reply:
Which begs another question, if result set iteration may issue pagination requests under the hood, won't it also block the thread?

Could the problem this PR fixes explain https://github.com/scylladb/scylladb/issues/14111 ?

Reply:
> If you drop all(), doesn't it mean that after the first iteration over the (paginated) result set in assert len(rows), the second loop for r in rows will be no-op? See https://github.com/scylladb/scylladb/commit/c39332710d3d0e4e17c879ea0b58550fecd12518

`run_async` is implemented by adding an `on_result` callback to the driver's `ResponseFuture` object; the result contains the whole list.
```python3
def _wrap_future(f: ResponseFuture) -> asyncio.Future:
    """Wrap a cassandra Future into an asyncio.Future object.

    Args:
        f: future to wrap

    Returns:
        And asyncio.Future object which can be awaited.
    """
    loop = asyncio.get_event_loop()
    aio_future = loop.create_future()

    def on_result(result):
        if not aio_future.done():
            loop.call_soon_threadsafe(aio_future.set_result, result)
        else:
            logger.debug("_wrap_future: on_result() on already done future: %s", result)

    def on_error(exception, *_):
        if not aio_future.done():
            loop.call_soon_threadsafe(aio_future.set_exception, exception)
        else:
            logger.debug("_wrap_future: on_error() on already done future: %s"), result

    f.add_callback(on_result)
    f.add_errback(on_error)
    return aio_future


def run_async(self, *args, **kwargs) -> asyncio.Future:
    # The default timeouts should have been more than enough, but in some
    # extreme cases with a very slow debug build running on a slow or very busy
    # machine, they may not be. Observed tests reach 160 seconds. So it's
    # incremented to 200 seconds.
    # See issue #11289.
    kwargs.setdefault("timeout", 200.0)
    return _wrap_future(self.execute_async(*args, **kwargs))
```

(this is not the smartest way to handle paging, but that's another issue...)

Reply:
In other words, `await cql.run_async(...)` already returns a `list`, not a generator or whatever.

Reply:
Well, since the failures happened in debug mode, there's another possible, very simple explanation.
Just read the comment in `run_async`.
```python3
    # The default timeouts should have been more than enough, but in some
    # extreme cases with a very slow debug build running on a slow or very busy
    # machine, they may not be. Observed tests reach 160 seconds. So it's
    # incremented to 200 seconds.
    # See issue #11289.
```

Reply:
Maybe we should add some guardrail to the test framework - if you try to use the blocking functions of the driver like `execute`, we'll immediately fail the test.